### PR TITLE
feat: support ARRAY_APPEND/ARRAY_REMOVE partial update ops for Array fields

### PR DESCRIPTION
**Pull request by weiliu1031**
issue: #49241
pr: https://github.com/milvus-io/milvus-proto/pull/586
design doc: https://github.com/milvus-io/milvus-design-docs/blob/main/design_docs/20260423-array_partial_update_ops.md

## Summary

Phase-1 implementation of field-level partial update operators for Array fields. Lets clients mutate a single array without resending the full value on upsert.

Supported ops (see `schemapb.FieldPartialUpdateOp.OpType`):
- `REPLACE` — legacy semantics (full overwrite). Default when no op is set.
- `ARRAY_APPEND` — concatenate payload to base (duplicates preserved, in-order).
- `ARRAY_REMOVE` — delete **all** occurrences matching any value in payload.

### Proto surface (via https://github.com/milvus-io/milvus-proto/pull/586)
- `UpsertRequest.field_ops` carries per-field directives; `FieldData` stays a pure data carrier (no leakage into Insert/Query/Search/msgstream).

### Proxy (`internal/proxy/task_upsert*.go`)
- `validateFieldPartialUpdateOps` rejects: empty field name, duplicate field, op on PK, unknown field, non-Array field, element-type mismatch, unsupported op enum, op without matching FieldData, ARRAY_APPEND payload + base > `max_capacity`.
- Any non-REPLACE op auto-promotes `partial_update=true` (user does not need to set both).
- In `queryPreExecute` merge step, dispatches per field: REPLACE → `UpdateFieldDataByColumn`; ARRAY_APPEND/REMOVE → `UpdateArrayFieldByColumnWithOp`.

### typeutil (`pkg/util/typeutil/schema.go`)
- `UpdateArrayFieldByColumnWithOp` (Array-only entry point) + `ApplyArrayRowOp` + `appendArrayRow` / `removeArrayRow` cover all Array element types (Bool/Int32/Int64/Float/Double/VarChar).
- REMOVE deletes every match (aligns with MongoDB `$pull`, Postgres `array_remove`).
- Float/Double REMOVE: `NaN != NaN` (IEEE 754).
- ARRAY_APPEND enforces `max_capacity` per merged row and returns a descriptive error.

### Go SDK (`client/v2`)
- `columnBasedDataOption.WithArrayAppend(field)` / `WithArrayRemove(field)` / `WithFieldPartialOp(field, op)` build `UpsertRequest.FieldOps`.
- REPLACE clears any prior directive for that field; non-REPLACE ops auto-promote `PartialUpdate=true`.
- Unknown-field ops are forwarded as-is so the server can return the descriptive validation error (client does not silently drop).

### Integration test (`tests/integration/hellomilvus/upsert_partial_op_test.go`)
End-to-end insert → op-upsert → query flow exercising: ARRAY_APPEND, ARRAY_REMOVE (incl. no-op row + multi-match), ARRAY_APPEND-over-capacity rejection. Does **not** set `partial_update=true` — asserts the server auto-promotes.

## Forward compatibility

Older servers that don't understand `field_ops` treat the request as a plain upsert (proto3 unknown-field tolerance) → graceful fallback to REPLACE semantics with a client-side warning path available.

## Test plan
- [x] typeutil UT: 40+ cases covering all element types, empty/nil inputs, capacity edge cases, NaN semantics, unsupported-type rejection
- [x] proxy UT: 20+ validation cases (PK, unknown field, type mismatch, capacity precheck, duplicate, empty name, op without FieldData)
- [x] go-sdk UT: 10 cases — auto-promote, REPLACE-clears-prior, row-based + column-based builders, multi-op emission
- [x] integration test: append / remove / capacity-overflow end-to-end
- [x] `go mod tidy` in root / pkg / client
- [ ] CI green

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: pkg/util/typeutil/schema.go [hunk 1]
```diff
@@ -1712,6 +1712,46 @@ func UpdateFieldData(base, update []*schemapb.FieldData, baseIdx, updateIdx int6
 	return nil
 }
 
+// UpdateArrayFieldByColumnWithOp merges an Array field's update rows into
+// the base rows while applying a FieldPartialUpdateOp. Non-Array field
+// types or a REPLACE op fall back to UpdateFieldDataByColumn's behavior.
+//
+// maxCapacity caps ARRAY_APPEND's post-merge length; pass -1 to skip the
+// check (proxy should pass the schema-declared max_capacity).
+func UpdateArrayFieldByColumnWithOp(
```
Comment (liliu-z) [inline pkg/util/typeutil/schema.go:1721]: UpdateArrayFieldByColumnWithOp does not synchronize ValidData, unlike UpdateFieldDataByColumn (L1763-1769). When a nullable Array row has ValidData[i]=false and an ARRAY_APPEND is applied, the merged data is silently treated as null downstream because ValidData is never flipped to true.

File: pkg/util/typeutil/schema.go [hunk 2]
```diff
@@ -3214,3 +3254,233 @@ func ExtractStructFieldName(fieldName string) (string, error) {
...
+	maxCapacity int,
+) (*schemapb.ScalarField, error) {
+	switch elementType {
+	case schemapb.DataType_Bool:
+		b := base.GetBoolData().GetData()
+		u := update.GetBoolData().GetData()
+		merged := make([]bool, 0, len(b)+len(u))
+		merged = append(merged, b...)
+		merged = append(merged, u...)
+		if maxCapacity >= 0 && len(merged) > maxCapacity {
```
Comment (liliu-z) [inline pkg/util/typeutil/schema.go:3312]: readMaxCapacity() returns 0 for missing/invalid capacity. The merge logic here treats 0 as a hard capacity limit and rejects appends, while the pre-check in task_upsert_partial_op.go:171 treats <=0 as 'skip check'. This inconsistency causes valid ARRAY_APPEND operations to be rejected when max_capacity is simply absent.

File: internal/proxy/task_upsert_partial_op.go [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+				return false, merr.WrapErrParameterInvalidMsg(
+					fmt.Sprintf("op %s requires Array field, but field %q is %s",
+						op.String(), name, fieldSchema.GetDataType().String()))
+			}
+		default:
+			return false, merr.WrapErrParameterInvalidMsg(
+				fmt.Sprintf("unsupported partial update op: %s", op.String()))
+		}
+
+		fd, ok := fieldDataByName[name]
```
Comment (liliu-z) [inline internal/proxy/task_upsert_partial_op.go:119]: No nil check on fd.GetScalars().GetArrayData() before passing to UpdateArrayFieldByColumnWithOp. A malformed FieldData with type=Array but nil ArrayData will cause a nil pointer panic at schema.go:1739.

**Comment by weiliu1031** [comment]
@liliu-z Thanks for the careful review. All three findings reproduced locally and fixed in 2446c7b34d.

1. **ValidData sync** (`schema.go:1721`) — merge loop now flips `base.ValidData[baseIdx]=true` on success and skips the row when `update.ValidData[updateIdx]=false`, so a null payload can never silently flip a row non-null, and a successful merge on a previously-null base is no longer dropped downstream.

2. **`readMaxCapacity` 0 vs -1 inconsistency** (`schema.go:3298` ↔ `task_upsert_partial_op.go:171`) — unified on **-1** as the "no capacity gate" sentinel. Pre-check uses `< 0` to skip, merge already honored `>= 0`. A schema with no declared max_capacity now correctly pass-through in both paths.

3. **nil ArrayData panic** (`task_upsert_partial_op.go:119`) — validate layer now rejects `Scalars==nil || ArrayData==nil` upfront with a `ParameterInvalid` error; also added a belt-and-suspenders nil guard at the merge entry point so any future caller can't trigger the deref.

Regression tests added:
- `TestUpdateArrayFieldByColumnWithOp_FlipsValidDataOnMerge`
- `TestUpdateArrayFieldByColumnWithOp_SkipsNullUpdatePayload`
- `TestUpdateArrayFieldByColumnWithOp_RejectsNilArrayData`
- `TestValidateFieldPartialUpdateOps_RejectsNilArrayData`
- `TestReadMaxCapacity` updated to assert the -1 sentinel.

PTAL.

**Comment by weiliu1031** [comment]
Rebased onto latest master (8fb3fef). Only one conflict in `client/milvusclient/write_options.go` — master added a `deferredErr` field to `columnBasedDataOption` and I had added `partialOps`; merged both cleanly. No logic changes during rebase.

Typeutil / client SDK UT green locally post-rebase. Proxy UT build currently fails on an unrelated C symbol (`SetArrowIOThreadPoolCapacity`, introduced by a recent master commit to `internal/util/initcore`) that requires a C++ core rebuild in my worktree; this has nothing to do with the partial-update changes. CI will run the full suite.
